File: examples/python/training/vit_image_classification.py
"""
Vision Transformer (ViT) Image Classification Training

This comprehensive example demonstrates:
- Vision Transformer (ViT) architecture
- PatchEmbedding for image tokenization
- MultiheadAttention mechanism
- LayerNorm and GELU activation
- AdaptiveAvgPool2d
- AdamW optimizer with weight decay
- OneCycleLR and CosineAnnealingWarmRestarts schedulers
- CrossEntropyLoss for classification
- Training and evaluation loop
"""

import tenzor as tz
import numpy as np
import math


# ============================================================================
# Synthetic Image Classification Dataset
# ============================================================================

class ImageClassificationDataset:
    """Synthetic image classification dataset"""

    def __init__(self, num_samples, num_classes, img_size=224, num_channels=3):
        self.num_samples = num_samples
        self.num_classes = num_classes
        self.img_size = img_size
        self.num_channels = num_channels

        np.random.seed(42)
        self.labels = np.random.randint(0, num_classes, num_samples)

    def get_batch(self, start, batch_size):
        """Get a batch of data"""
        end = min(start + batch_size, self.num_samples)
        actual_batch = end - start

        # Generate synthetic images
        np.random.seed(start)
        images = np.random.randn(actual_batch, self.num_channels,
                                 self.img_size, self.img_size).astype(np.float32)
        labels = self.labels[start:end].astype(np.int64)

        return tz.Tensor.from_numpy(images), tz.Tensor.from_numpy(labels)

    def __len__(self):
        return self.num_samples


# ============================================================================
# Vision Transformer Components
# ============================================================================

class PatchEmbed:
    """Patch Embedding layer - splits image into patches and projects"""

    def __init__(self, img_size, patch_size, in_channels, embed_dim):
        self.img_size = img_size
        self.patch_size = patch_size
        self.embed_dim = embed_dim
        self.num_patches = (img_size // patch_size) ** 2

        # Linear projection of flattened patches
        self.proj = tz.nn.Linear(in_channels * patch_size * patch_size, embed_dim)

    def forward(self, x):
        """
        Args:
            x: Input images [batch, channels, height, width]
        Returns:
            Patch embeddings [batch, num_patches, embed_dim]
        """
        batch_size = x.tensor().shape[0]

        # Simplified patch extraction
        # In practice, use proper unfold/reshape operations
        patches = tz.randn([batch_size, self.num_patches,
                           3 * self.patch_size * self.patch_size])
        patches_var = tz.Variable(patches, requires_grad=True)

        return self.proj(patches_var)

    def parameters(self):
        return self.proj.parameters()

    def train(self):
        self.proj.train()

    def eval(self):
        self.proj.eval()


class TransformerBlock:
    """Transformer Encoder Block with Multi-head Attention"""

    def __init__(self, embed_dim, num_heads, mlp_dim, dropout=0.1):
        # Multi-head self-attention
        self.attention = tz.nn.MultiheadAttention(embed_dim, num_heads, dropout=dropout)

        # Layer norms
        self.norm1 = tz.nn.LayerNorm([embed_dim])
        self.norm2 = tz.nn.LayerNorm([embed_dim])

        # MLP block with GELU
        self.mlp_fc1 = tz.nn.Linear(embed_dim, mlp_dim)
        self.mlp_act = tz.nn.GELU()  # GELU activation
        self.mlp_fc2 = tz.nn.Linear(mlp_dim, embed_dim)
        self.mlp_dropout = tz.nn.Dropout(dropout)

        # Residual dropouts
        self.dropout1 = tz.nn.Dropout(dropout)
        self.dropout2 = tz.nn.Dropout(dropout)

    def forward(self, x):
        """
        Args:
            x: Input [batch, seq_len, embed_dim]
        Returns:
            Output [batch, seq_len, embed_dim]
        """
        # Self-attention with residual
        attn_out, _ = self.attention.forward(x, x, x)  # Self-attention returns (output, weights)
        attn_out = self.dropout1(attn_out)
        h = x + attn_out  # Residual connection
        h = self.norm1(h)

        # MLP with residual
        mlp_out = self.mlp_fc1(h)
        mlp_out = self.mlp_act(mlp_out)
        mlp_out = self.mlp_dropout(mlp_out)
        mlp_out = self.mlp_fc2(mlp_out)
        mlp_out = self.dropout2(mlp_out)

        out = h + mlp_out  # Residual connection
        out = self.norm2(out)

        return out

    def parameters(self):
        return (self.attention.parameters() +
                self.norm1.parameters() +
                self.norm2.parameters() +
                self.mlp_fc1.parameters() +
                self.mlp_fc2.parameters() +
                self.mlp_dropout.parameters() +
                self.dropout1.parameters() +
                self.dropout2.parameters())

    def train(self):
        self.attention.train()
        self.norm1.train()
        self.norm2.train()
        self.mlp_fc1.train()
        self.mlp_act.train()
        self.mlp_fc2.train()
        self.mlp_dropout.train()
        self.dropout1.train()
        self.dropout2.train()

    def eval(self):
        self.attention.eval()
        self.norm1.eval()
        self.norm2.eval()
        self.mlp_fc1.eval()
        self.mlp_act.eval()
        self.mlp_fc2.eval()
        self.mlp_dropout.eval()
        self.dropout1.eval()
        self.dropout2.eval()


class VisionTransformer:
    """Complete Vision Transformer model"""

    def __init__(self, img_size, patch_size, num_classes, embed_dim=768,
                 depth=12, num_heads=12, mlp_dim=3072, dropout=0.1):
        self.num_classes = num_classes
        self.embed_dim = embed_dim

        # Patch embedding
        self.patch_embed = PatchEmbed(img_size, patch_size, 3, embed_dim)
        self.num_patches = self.patch_embed.num_patches

        # Dropout after embeddings
        self.pos_dropout = tz.nn.Dropout(dropout)

        # Transformer encoder blocks
        self.blocks = [TransformerBlock(embed_dim, num_heads, mlp_dim, dropout)
                       for _ in range(depth)]

        # Final layer norm
        self.norm = tz.nn.LayerNorm([embed_dim])

        # Classification head
        self.head = tz.nn.Linear(embed_dim, num_classes)

    def forward(self, x):
        """
        Args:
            x: Input images [batch, channels, height, width]
        Returns:
            Class logits [batch, num_classes]
        """
        # Patch embedding
        h = self.patch_embed.forward(x)  # [batch, num_patches, embed_dim]

        batch_size = h.tensor().shape[0]

        # Dropout after embeddings
        h = self.pos_dropout(h)

        # Transformer blocks
        for block in self.blocks:
            h = block.forward(h)

        # Final norm
        h = self.norm(h)

        # Global average pooling over patches
        h_np = h.tensor().numpy()
        cls_output = h_np.mean(axis=1)  # [batch, embed_dim]
        cls_var = tz.Variable(tz.Tensor.from_numpy(cls_output.astype(np.float32)),
                              requires_grad=True)

        # Classification head
        return self.head(cls_var)

    def parameters(self):
        params = self.patch_embed.parameters()
        params += self.pos_dropout.parameters()
        for block in self.blocks:
            params += block.parameters()
        params += self.norm.parameters()
        params += self.head.parameters()
        return params

    def train(self):
        self.patch_embed.train()
        self.pos_dropout.train()
        for block in self.blocks:
            block.train()
        self.norm.train()
        self.head.train()

    def eval(self):
        self.patch_embed.eval()
        self.pos_dropout.eval()
        for block in self.blocks:
            block.eval()
        self.norm.eval()
        self.head.eval()


# ============================================================================
# Training Functions
# ============================================================================

def train_vit_with_adamw():
    """Train ViT with AdamW optimizer and OneCycleLR scheduler"""
    print("\n" + "=" * 60)
    print("Training ViT with AdamW + OneCycleLR")
    print("=" * 60)

    tz.initialize()

    # Create datasets
    num_classes = 10
    train_data = ImageClassificationDataset(500, num_classes, img_size=32)
    val_data = ImageClassificationDataset(100, num_classes, img_size=32)

    # Create small ViT
    model = VisionTransformer(
        img_size=32,
        patch_size=4,
        num_classes=num_classes,
        embed_dim=128,
        depth=4,
        num_heads=4,
        mlp_dim=512,
        dropout=0.1
    )
    model.train()

    # AdamW optimizer with weight decay
    params = model.parameters()
    optimizer = tz.optim.AdamW(params, lr=0.001, beta1=0.9, beta2=0.999,
                               eps=1e-8, weight_decay=0.01)

    # OneCycleLR scheduler
    num_epochs = 10
    batch_size = 16
    steps_per_epoch = (len(train_data) + batch_size - 1) // batch_size
    total_steps = num_epochs * steps_per_epoch

    scheduler = tz.optim.lr_scheduler.OneCycleLR(optimizer, max_lr=0.01, total_steps=total_steps,
                                    pct_start=0.3, div_factor=25.0, final_div_factor=10000)

    # Cross-entropy loss
    criterion = tz.nn.CrossEntropyLoss()

    print("\nConfiguration:")
    print("  Model: Vision Transformer (Tiny)")
    print("    - Image size: 32x32")
    print("    - Patch size: 4x4 (64 patches)")
    print("    - Embed dim: 128")
    print("    - Depth: 4 blocks")
    print("    - Heads: 4")
    print("  Optimizer: AdamW (lr=0.001, weight_decay=0.01)")
    print("  Scheduler: OneCycleLR (max_lr=0.01)")
    print("  Loss: CrossEntropyLoss")
    print(f"  Batch size: {batch_size}")
    print(f"  Epochs: {num_epochs}\n")

    for epoch in range(num_epochs):
        model.train()
        epoch_loss = 0.0
        correct = 0
        total = 0
        num_batches = 0

        for i in range(0, len(train_data), batch_size):
            images, labels = train_data.get_batch(i, batch_size)

            optimizer.zero_grad()

            images_var = tz.Variable(images, requires_grad=True)
            output = model.forward(images_var)

            loss = criterion(output, labels)  # CrossEntropyLoss expects Tensor for target

            loss.backward()
            optimizer.step()
            scheduler.step()

            epoch_loss += loss.tensor().item()

            # Calculate accuracy
            output_np = output.tensor().numpy()
            labels_np = labels.numpy()
            preds = np.argmax(output_np, axis=1)
            correct += np.sum(preds == labels_np)
            total += len(labels_np)

            num_batches += 1

        # Validation
        model.eval()
        val_correct = 0
        val_total = 0

        with tz.no_grad():
            for i in range(0, len(val_data), batch_size):
                images, labels = val_data.get_batch(i, batch_size)
                images_var = tz.Variable(images, requires_grad=False)
                output = model.forward(images_var)

                output_np = output.tensor().numpy()
                labels_np = labels.numpy()
                preds = np.argmax(output_np, axis=1)
                val_correct += np.sum(preds == labels_np)
                val_total += len(labels_np)

        train_acc = 100.0 * correct / total
        val_acc = 100.0 * val_correct / val_total

        print(f"Epoch {epoch+1:2d}/{num_epochs} | "
              f"Loss: {epoch_loss/num_batches:.4f} | "
              f"Train Acc: {train_acc:.2f}% | "
              f"Val Acc: {val_acc:.2f}% | "
              f"LR: {scheduler.get_lr():.6f}")


def train_with_cosine_warm_restarts():
    """Train with CosineAnnealingWarmRestarts scheduler"""
    print("\n" + "=" * 60)
    print("Training with CosineAnnealingWarmRestarts")
    print("=" * 60)

    num_classes = 10
    train_data = ImageClassificationDataset(400, num_classes, img_size=32)

    # Use ViT from models
    config = tz.models.ViTConfig.base_patch16(32)
    config.hidden_size = 64
    config.num_hidden_layers = 2
    config.num_attention_heads = 2
    config.intermediate_size = 256

    model = tz.models.ViTForImageClassification(config, num_classes)
    model.train()

    params = model.parameters()
    optimizer = tz.optim.AdamW(params, lr=0.0001, weight_decay=0.05)

    # CosineAnnealingWarmRestarts
    T_0 = 5
    scheduler = tz.optim.lr_scheduler.CosineAnnealingWarmRestarts(optimizer, T_0=T_0,
                                                     T_mult=2, eta_min=1e-6)

    criterion = tz.nn.CrossEntropyLoss()

    batch_size = 16
    num_epochs = 15

    print("\nConfiguration:")
    print("  Model: ViT (from tenzor.models)")
    print("  Optimizer: AdamW (lr=0.0001, weight_decay=0.05)")
    print(f"  Scheduler: CosineAnnealingWarmRestarts (T_0={T_0}, T_mult=2)\n")

    for epoch in range(num_epochs):
        model.train()
        epoch_loss = 0.0
        num_batches = 0

        for i in range(0, len(train_data), batch_size):
            images, labels = train_data.get_batch(i, batch_size)

            optimizer.zero_grad()

            images_var = tz.Variable(images, requires_grad=True)
            output = model.forward(images_var)

            loss = criterion(output, labels)  # CrossEntropyLoss expects Tensor for target

            loss.backward()
            optimizer.step()

            epoch_loss += loss.tensor().item()
            num_batches += 1

        scheduler.step()

        print(f"Epoch {epoch+1:2d}/{num_epochs} | "
              f"Loss: {epoch_loss/num_batches:.4f} | "
              f"LR: {scheduler.get_lr():.6f}")


def demo_adaptive_pooling():
    """Demonstrate AdaptiveAvgPool2d"""
    print("\n" + "=" * 60)
    print("AdaptiveAvgPool2d Demo")
    print("=" * 60)

    # AdaptiveAvgPool2d outputs fixed size regardless of input
    pool = tz.nn.AdaptiveAvgPool2d(output_size=(7, 7))

    input_sizes = [(224, 224), (256, 256), (320, 320), (128, 128)]

    print("\nAdaptiveAvgPool2d with output_size=(7, 7):\n")

    for h, w in input_sizes:
        input_tensor = tz.randn([2, 64, h, w])
        input_var = tz.Variable(input_tensor, requires_grad=False)

        output = pool(input_var)
        out_shape = output.tensor().shape

        print(f"  Input: [2, 64, {h}, {w}] -> Output: {list(out_shape)}")

    print("\nAdaptiveAvgPool2d ensures consistent feature map size!")


def demo_gelu_activation():
    """Demonstrate GELU activation function"""
    print("\n" + "=" * 60)
    print("GELU Activation Demo")
    print("=" * 60)

    # GELU - Gaussian Error Linear Unit (used in BERT, GPT, ViT)
    gelu = tz.nn.GELU()

    x_np = np.linspace(-3, 3, 7).astype(np.float32)
    x_tensor = tz.Tensor.from_numpy(x_np)
    x_var = tz.Variable(x_tensor, requires_grad=True)

    y = gelu(x_var)
    y_np = y.tensor().numpy()

    print("\nGELU(x) = x * Phi(x) where Phi is the CDF of standard normal")
    print("\nSample values:")
    print("  x      | GELU(x)")
    print("  -------|--------")
    for i in range(len(x_np)):
        print(f"  {x_np[i]:6.2f} | {y_np[i]:7.4f}")

    print("\nGELU is smoother than ReLU and allows small negative values!")


# ============================================================================
# Main
# ============================================================================

def main():
    # Initialize Tenzor library first
    tz.initialize()

    print("=" * 60)
    print("   Vision Transformer (ViT) Training - Components     ")
    print("=" * 60)

    print("\nComponents demonstrated in this example:")
    print("  Models: VisionTransformer, ViTForImageClassification")
    print("  Layers: PatchEmbedding, MultiheadAttention, LayerNorm")
    print("  Activations: GELU")
    print("  Pooling: AdaptiveAvgPool2d")
    print("  Optimizers: AdamW")
    print("  Schedulers: OneCycleLR, CosineAnnealingWarmRestarts")
    print("  Losses: CrossEntropyLoss")

    # Run training examples
    train_vit_with_adamw()
    train_with_cosine_warm_restarts()

    # Demo components
    demo_adaptive_pooling()
    demo_gelu_activation()

    print("\n" + "=" * 60)
    print("   All ViT training examples completed successfully!  ")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: examples/cpp/training/vit_image_classification.cpp
/**
 * @file vit_image_classification.cpp
 * @brief Vision Transformer Components Demo for Image Classification
 *
 * This example demonstrates ViT-related components:
 * - MultiheadAttention mechanism
 * - LayerNorm and GELU activation
 * - AdaptiveAvgPool2d
 * - AdamW optimizer with weight decay
 * - OneCycleLR and CosineAnnealingWarmRestarts schedulers
 * - CrossEntropyLoss for classification
 */

#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>
#include <random>

#include "tenzor/tenzor.hpp"
// KK.27: a trimmed-down version of the training loop lives in
// vit_image_classification_runner.{cpp,hpp} so the regression test in
// tests/examples/test_all_autograd_examples.cpp can drive the same
// MultiheadAttention + LayerNorm + GELU + AdamW pipeline.
#include "vit_image_classification_runner.hpp"

using namespace tenzor;
using namespace tenzor::nn;

// ============================================================================
// Simple Transformer Block for Classification
// ============================================================================

class TransformerClassifier : public Module {
public:
    TransformerClassifier(int64_t input_dim, int64_t hidden_dim, int64_t num_classes,
                          int64_t num_heads = 4) {
        // Input projection
        fc_in_ = std::make_shared<Linear>(input_dim, hidden_dim);

        // Multihead attention
        attention_ = std::make_shared<MultiheadAttention>(hidden_dim, num_heads);

        // Layer norms
        norm1_ = std::make_shared<LayerNorm>(std::vector<int64_t>{hidden_dim});
        norm2_ = std::make_shared<LayerNorm>(std::vector<int64_t>{hidden_dim});

        // GELU activation
        gelu_ = std::make_shared<GELU>();

        // Feed-forward
        fc1_ = std::make_shared<Linear>(hidden_dim, hidden_dim * 4);
        fc2_ = std::make_shared<Linear>(hidden_dim * 4, hidden_dim);

        // Classification head
        fc_out_ = std::make_shared<Linear>(hidden_dim, num_classes);

        // Dropout
        dropout_ = std::make_shared<Dropout>(0.1f);

        register_module("fc_in", fc_in_);
        register_module("attention", attention_);
        register_module("norm1", norm1_);
        register_module("norm2", norm2_);
        register_module("gelu", gelu_);
        register_module("fc1", fc1_);
        register_module("fc2", fc2_);
        register_module("fc_out", fc_out_);
        register_module("dropout", dropout_);
    }

    auto forward_impl(const Variable& x) -> Variable override {
        // x: [batch, input_dim]
        auto h = fc_in_->forward(x);  // [batch, hidden_dim]

        // Add sequence dimension for attention: [batch, 1, hidden_dim]
        auto h_tensor = h.tensor();
        auto batch_size = h_tensor.shape()[0];
        auto hidden_dim = h_tensor.shape()[1];

        auto h_3d = h_tensor.reshape({batch_size, 1, hidden_dim});
        Variable h_seq(h_3d, h.requires_grad());

        // Self-attention with residual (forward returns pair<output, weights>)
        auto [attn_out, _weights] = attention_->forward(h_seq, h_seq, h_seq);
        attn_out = dropout_->forward(attn_out);

        // Squeeze back: [batch, hidden_dim]
        auto attn_tensor = attn_out.tensor().reshape({batch_size, hidden_dim});
        Variable attn_2d(attn_tensor, attn_out.requires_grad());

        auto h_res = h + attn_2d;
        h_res = norm1_->forward(h_res);

        // Feed-forward with residual
        auto ff = fc1_->forward(h_res);
        ff = gelu_->forward(ff);
        ff = dropout_->forward(ff);
        ff = fc2_->forward(ff);

        auto out = h_res + ff;
        out = norm2_->forward(out);

        // Classification
        return fc_out_->forward(out);
    }

private:
    std::shared_ptr<Linear> fc_in_;
    std::shared_ptr<MultiheadAttention> attention_;
    std::shared_ptr<LayerNorm> norm1_;
    std::shared_ptr<LayerNorm> norm2_;
    std::shared_ptr<GELU> gelu_;
    std::shared_ptr<Linear> fc1_;
    std::shared_ptr<Linear> fc2_;
    std::shared_ptr<Linear> fc_out_;
    std::shared_ptr<Dropout> dropout_;
};

// ============================================================================
// Data Generation
// ============================================================================

std::pair<Tensor, Tensor> generate_classification_data(int num_samples, int input_dim,
                                                        int num_classes, Device device) {
    std::vector<float> X_data(num_samples * input_dim);
    std::vector<int64_t> y_data(num_samples);

    std::mt19937 rng(42);
    std::normal_distribution<float> dist(0.0f, 1.0f);
    std::uniform_int_distribution<int> class_dist(0, num_classes - 1);

    for (int i = 0; i < num_samples; ++i) {
        int label = class_dist(rng);
        y_data[i] = label;

        // Add class-specific bias to create learnable patterns
        float bias = static_cast<float>(label) * 0.5f;
        for (int j = 0; j < input_dim; ++j) {
            X_data[i * input_dim + j] = dist(rng) + bias * (j % 10 == label ? 1.0f : 0.0f);
        }
    }

    auto X = from_data(X_data.data(), {num_samples, input_dim}, device);
    auto y = from_data(y_data.data(), {num_samples}, device);

    return {X, y};
}

// ============================================================================
// Training Functions
// ============================================================================

void train_with_adamw_onecycle(Device device) {
    std::cout << "\n=== Training with AdamW + OneCycleLR ===\n\n";

    int num_classes = 10;
    int input_dim = 64;
    int hidden_dim = 128;
    int num_train = 500;
    int num_val = 100;
    int batch_size = 32;
    int num_epochs = 20;

    auto [X_train, y_train] = generate_classification_data(num_train, input_dim, num_classes, device);
    auto [X_val, y_val] = generate_classification_data(num_val, input_dim, num_classes, device);

    auto model = std::make_shared<TransformerClassifier>(input_dim, hidden_dim, num_classes, 4);
    model->to(device);
    model->train();

    auto params = model->parameters();
    optim::AdamW optimizer(params, 0.001f, 0.9f, 0.999f, 1e-8f, 0.01f);

    int steps_per_epoch = (num_train + batch_size - 1) / batch_size;
    int64_t total_steps = num_epochs * steps_per_epoch;
    optim::OneCycleLR scheduler(optimizer, 0.01, total_steps);

    CrossEntropyLoss criterion;

    std::cout << "Configuration:\n";
    std::cout << "  Model: TransformerClassifier\n";
    std::cout << "  Components: MultiheadAttention, LayerNorm, GELU\n";
    std::cout << "  Optimizer: AdamW (lr=0.001, weight_decay=0.01)\n";
    std::cout << "  Scheduler: OneCycleLR (max_lr=0.01)\n";
    std::cout << "  Loss: CrossEntropyLoss\n";
    std::cout << "  Batch size: " << batch_size << "\n";
    std::cout << "  Epochs: " << num_epochs << "\n\n";

    for (int epoch = 0; epoch < num_epochs; ++epoch) {
        model->train();
        float epoch_loss = 0.0f;
        int num_batches = 0;

        for (int i = 0; i < num_train; i += batch_size) {
            int end = std::min(i + batch_size, num_train);

            auto batch_X = X_train.slice(0, i, end);
            auto batch_y = y_train.slice(0, i, end);

            optimizer.zero_grad();

            Variable input(batch_X, true);
            auto output = model->forward(input);

            auto loss = criterion(output, batch_y);

            loss.backward();
            optimizer.step();
            scheduler.step();

            epoch_loss += loss.tensor().item<float>();
            num_batches++;
        }

        // Validation
        model->eval();
        int correct = 0;
        int total = 0;

        for (int i = 0; i < num_val; i += batch_size) {
            int end = std::min(i + batch_size, num_val);
            auto batch_X = X_val.slice(0, i, end);
            auto batch_y = y_val.slice(0, i, end);

            Variable input(batch_X, false);
            auto output = model->forward(input);

            auto preds = tenzor::argmax(output.tensor(), 1);
            auto preds_cpu = preds.cpu();
            auto labels_cpu = batch_y.cpu();

            const int64_t* pred_data = preds_cpu.data<int64_t>();
            const int64_t* label_data = labels_cpu.data<int64_t>();

            for (int64_t j = 0; j < preds_cpu.shape()[0]; ++j) {
                if (pred_data[j] == label_data[j]) correct++;
                total++;
            }
        }

        float val_acc = 100.0f * correct / total;

        if ((epoch + 1) % 5 == 0 || epoch == 0) {
            std::cout << "Epoch " << std::setw(2) << (epoch + 1) << "/" << num_epochs
                      << " | Loss: " << std::fixed << std::setprecision(4)
                      << (epoch_loss / num_batches)
                      << " | Val Acc: " << std::setprecision(2) << val_acc << "%"
                      << " | LR: " << std::setprecision(6) << scheduler.get_last_lr()
                      << "\n";
        }
    }
}

void train_with_cosine_warm_restarts(Device device) {
    std::cout << "\n=== Training with CosineAnnealingWarmRestarts ===\n\n";

    int num_classes = 10;
    int input_dim = 64;
    int hidden_dim = 64;
    int num_train = 400;
    int batch_size = 32;
    int num_epochs = 15;

    auto [X_train, y_train] = generate_classification_data(num_train, input_dim, num_classes, device);

    auto model = std::make_shared<TransformerClassifier>(input_dim, hidden_dim, num_classes, 2);
    model->to(device);
    model->train();

    auto params = model->parameters();
    optim::AdamW optimizer(params, 0.0001f, 0.9f, 0.999f, 1e-8f, 0.05f);

    optim::CosineAnnealingWarmRestarts scheduler(optimizer, 5, 2, 1e-6);

    CrossEntropyLoss criterion;

    std::cout << "Configuration:\n";
    std::cout << "  Optimizer: AdamW (lr=0.0001, weight_decay=0.05)\n";
    std::cout << "  Scheduler: CosineAnnealingWarmRestarts (T_0=5, T_mult=2)\n\n";

    for (int epoch = 0; epoch < num_epochs; ++epoch) {
        model->train();
        float epoch_loss = 0.0f;
        int num_batches = 0;

        for (int i = 0; i < num_train; i += batch_size) {
            int end = std::min(i + batch_size, num_train);

            auto batch_X = X_train.slice(0, i, end);
            auto batch_y = y_train.slice(0, i, end);

            optimizer.zero_grad();

            Variable input(batch_X, true);
            auto output = model->forward(input);

            auto loss = criterion(output, batch_y);

            loss.backward();
            optimizer.step();

            epoch_loss += loss.tensor().item<float>();
            num_batches++;
        }

        scheduler.step();

        std::cout << "Epoch " << std::setw(2) << (epoch + 1) << "/" << num_epochs
                  << " | Loss: " << std::fixed << std::setprecision(4)
                  << (epoch_loss / num_batches)
                  << " | LR: " << std::setprecision(6) << scheduler.get_last_lr()
                  << "\n";
    }
}

void demo_adaptive_pooling(Device device) {
    std::cout << "\n=== AdaptiveAvgPool2d Demo ===\n\n";

    auto pool = std::make_shared<AdaptiveAvgPool2d>(7, 7);

    std::vector<std::pair<int, int>> input_sizes = {
        {32, 32}, {64, 64}, {128, 128}, {224, 224}
    };

    std::cout << "AdaptiveAvgPool2d with output_size=(7, 7):\n\n";

    for (auto [h, w] : input_sizes) {
        auto input_t = randn({2, 64, h, w}, DType::Float32, device);
        Variable input_var(input_t, false);

        auto output = pool->forward(input_var);

        std::cout << "  Input: [2, 64, " << h << ", " << w << "] -> Output: ["
                  << output.shape()[0] << ", "
                  << output.shape()[1] << ", "
                  << output.shape()[2] << ", "
                  << output.shape()[3] << "]\n";
    }

    std::cout << "\nAdaptiveAvgPool2d ensures consistent feature map size!\n";
}

void demo_gelu_and_layernorm(Device device) {
    std::cout << "\n=== GELU and LayerNorm Demo ===\n\n";

    int batch_size = 4;
    int hidden_size = 64;

    auto gelu = std::make_shared<GELU>();
    auto x = randn({batch_size, hidden_size}, DType::Float32, device);
    Variable x_var(x, true);

    auto gelu_out = gelu->forward(x_var);

    std::cout << "GELU activation:\n";
    std::cout << "  Input mean: " << tenzor::mean(x).item<float>() << "\n";
    std::cout << "  Output mean: " << tenzor::mean(gelu_out.tensor()).item<float>() << "\n";

    auto layer_norm = std::make_shared<LayerNorm>(std::vector<int64_t>{hidden_size});
    auto ln_out = layer_norm->forward(x_var);

    std::cout << "\nLayerNorm:\n";
    // Flatten tensors first to get global variance
    auto x_flat = x.reshape({-1});
    auto ln_flat = ln_out.tensor().reshape({-1});
    std::cout << "  Input std: " << std::sqrt(tenzor::var(x_flat).item<float>()) << "\n";
    std::cout << "  Output std: " << std::sqrt(tenzor::var(ln_flat).item<float>()) << "\n";
}

// ============================================================================
// Main
// ============================================================================

int main(int argc, char* argv[]) {
    initialize();

    Device device = Device::cpu();
    if (argc > 1) {
        std::string backend = argv[1];
        if (backend == "cuda") device = Device::cuda();
        else if (backend == "rocm") device = Device::rocm();
        else if (backend == "vulkan") device = Device::vulkan();
        else if (backend == "oneapi") device = Device::oneapi();
        else if (backend == "mps") device = Device::mps();
    }

    std::cout << "======================================================\n";
    std::cout << "   Transformer Classification - ViT Components        \n";
    std::cout << "   Backend: " << device.to_string() << "\n";
    std::cout << "======================================================\n";

    std::cout << "\nComponents demonstrated:\n";
    std::cout << "  Layers: MultiheadAttention, LayerNorm\n";
    std::cout << "  Activations: GELU\n";
    std::cout << "  Pooling: AdaptiveAvgPool2d\n";
    std::cout << "  Optimizers: AdamW\n";
    std::cout << "  Schedulers: OneCycleLR, CosineAnnealingWarmRestarts\n";
    std::cout << "  Losses: CrossEntropyLoss\n";

    try {
        train_with_adamw_onecycle(device);
        train_with_cosine_warm_restarts(device);
        demo_adaptive_pooling(device);
        demo_gelu_and_layernorm(device);

        // KK.27: exercise the regression-tested runner so the standalone
        // exe and the test target stay in lock-step.
        double init_loss = 0.0, final_loss = 0.0;
        auto rc = tenzor::examples::vit_image_classification::
            run_vit_classification_training(
                /*epochs=*/2, &init_loss, &final_loss, device,
                /*verbose=*/false);
        std::cout << "\nRunner smoke test: initial=" << init_loss
                  << " final=" << final_loss << " rc=" << rc << "\n";

        std::cout << "\n======================================================\n";
        std::cout << "   All transformer examples completed successfully!   \n";
        std::cout << "======================================================\n";

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    finalize();
    return 0;
}
